database/tests/desktop/core/sync_point_spec_test.cc

```diff
@@ -0,0 +1,883 @@
+    EXPECT_TRUE(false) << "Can't raise test events!";
+  }
+
+  void FireCancelEvent(Error error) override {
+    EXPECT_TRUE(false) << "Can't raise test events!";
+  }
+
+  bool MatchesListener(const void* listener_ptr) const override {
+    EXPECT_TRUE(false) << "Can't raise test events!";
+    return static_cast<const void*>(this) == listener_ptr;
+  }
+};
+
+UniquePtr<QueryInternal> ParseQuery(
+    UniquePtr<QueryInternal> query,
+    const test_data::QueryParams* query_params) {
+  EXPECT_NE(query_params->tag(), 0) << "Non-default queries must have a tag";
+  if (query_params->orderBy()) {
+    query.reset(query->OrderByChild(query_params->orderBy()->c_str()));
+  } else if (query_params->orderByKey()) {
+    query.reset(query->OrderByKey());
+  } else if (query_params->orderByPriority()) {
+    query.reset(query->OrderByPriority());
+  }
+  if (query_params->startAt()) {
+    const test_data::Bound* bound = query_params->startAt();
+    const char* name = bound->name() ? bound->name()->c_str() : "";
+    Variant index =
+        bound->index()
+            ? util::FlexbufferToVariant(bound->index_flexbuffer_root())
+            : Variant::Null();
+    query.reset(query->StartAt(index, name));
+  }
+  if (query_params->endAt()) {
+    const test_data::Bound* bound = query_params->endAt();
+    const char* name = bound->name() ? bound->name()->c_str() : "";
+    Variant index =
+        bound->index()
+            ? util::FlexbufferToVariant(bound->index_flexbuffer_root())
+            : Variant::Null();
+    query.reset(query->EndAt(index, name));
+  }
+  if (query_params->equalTo()) {
+    const test_data::Bound* bound = query_params->equalTo();
+    const char* name = bound->name() ? bound->name()->c_str() : "";
+    Variant index =
+        bound->index()
+            ? util::FlexbufferToVariant(bound->index_flexbuffer_root())
+            : Variant::Null();
+    query.reset(query->EqualTo(index, name));
+  }
+  if (query_params->limitToFirst()) {
+    query.reset(query->LimitToFirst(query_params->limitToFirst()));
+  }
+  if (query_params->limitToLast()) {
+    query.reset(query->LimitToLast(query_params->limitToLast()));
+  }
+  return query;
+}
+
+Event ParseEvent(const test_data::Event* event_spec, Path path) {
+  EventType type = static_cast<EventType>(event_spec->type());
+
+  EXPECT_NE(event_spec->path(), nullptr);
+  path = path.GetChild(event_spec->path()->c_str());
+  if (event_spec->name() != nullptr) {
+    path = path.GetChild(event_spec->name()->c_str());
+  }
+  Variant data = event_spec->data()
+                     ? FlexbufferToVariant(event_spec->data_flexbuffer_root())
+                     : Variant::Null();
+  DataSnapshotInternal snapshot(nullptr, data, QuerySpec(path));
+
+  const char* prev_name =
+      event_spec->prevName() != nullptr ? event_spec->prevName()->c_str() : "";
+  return Event(type, nullptr, snapshot, prev_name);
+}
+
+void SyncTreeTest::runTest(const test_data::TestCase* test_spec,
```

Comment:
Is it possible to give any more context in the expect/assert output to help narrow down the issue if there's a failure?
For example, EXPECT_THAT(...) << "Listen step failed in " << test_spec->name() << ...any additional info

Reply:
The asserts in the library are pretty good about spitting out an error message saying why they're crashing specifically. While the tests have names, the individual steps do not. However, the tests that do interesting things have a "comment" field that describes what they are doing, so if you do get a failure you can see what the most recent Test and comment are. Additionally, the EXPECT_THAT(..., Pointwise(...)) construction will tell you whether there are elements missing, or elements that aren't the expected value, and at what index the element appears, etc.

I can add the step index if you think it would be helpful, but I'm not convinced it would speed debugging up any.

I do have a commit stashed in my local repo that contains a bunch of `operator<<` functions for various classes that print out the values of the class, I can check these in if you'd like. This can help with printing out error messages when, for example, an event isn't the expected value.

Reply:
I'd definitely err on the side of giving more info, but if you want to do that in a follow-up that's fine.